llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp

```diff
@@ -1202,15 +1208,14 @@ bool RISCVVLOptimizer::isCandidate(const MachineInstr &MI) const {
   return true;
 }
 
-std::optional<MachineOperand>
-RISCVVLOptimizer::getMinimumVLForUser(MachineOperand &UserOp) {
+MachineOperand RISCVVLOptimizer::getMinimumVLForUser(MachineOperand &UserOp) {
   const MachineInstr &UserMI = *UserOp.getParent();
   const MCInstrDesc &Desc = UserMI.getDesc();
 
   if (!RISCVII::hasVLOp(Desc.TSFlags) || !RISCVII::hasSEWOp(Desc.TSFlags)) {
     LLVM_DEBUG(dbgs() << "    Abort due to lack of VL, assume that"
                          " use VLMAX\n");
-    return std::nullopt;
+    return MachineOperand::CreateImm(RISCV::VLMaxSentinel);
```

Comment:
Not related to this patch, but I wonder if we should avoid calling this function when a user does not have a VL operand. The debug output says that assume it uses VLMAX, but I wonder if it just doesn't depend on VL and could have been excluded from calculating the demanded VL.

Reply:
If a user doesn't have a VL operand it could be a non-pseudo instruction like a COPY etc, which in that case I think we need to preserve the entire VL? I.e. it would still demand VLMAX.

Reply:
I think you are right. Imagine we return the DEST of the copy, then we need all lanes of the copy. Please ignore my original comment. It does make me wonder if there is any future work related to the VL optimizer looking through copies though.

Reply:
I actually have a patch that deals with copies of V0 that I hope to post soon. 

We currently can't propagate VL through any mask uses because they have to be copied to a physical register first, so my solution was to keep track of all the uses of V0 in a map, similar to how we do it for defs in RISCVVectorPeephole